debugtools/DDR_VM/src/com/ibm/j9ddr/tools/ddrinteractive/CacheMemorySource.java

```diff
@@ -0,0 +1,133 @@
+ * General Public License, version 2 with the GNU Classpath                                                                                                                  
+ * Exception [1] and GNU General Public License, version 2 with the                                                                                                          
+ * OpenJDK Assembly Exception [2].                                                                                                                                           
+ *                                                                                                                                                                           
+ * [1] https://www.gnu.org/software/classpath/license.html                                                                                                                   
+ * [2] http://openjdk.java.net/legal/assembly-exception.html                                                                                                                 
+ *                                                                                                                                                                           
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception                                       
+ *******************************************************************************/
+package com.ibm.j9ddr.tools.ddrinteractive;
+
+import java.nio.ByteOrder;
+import java.nio.ByteBuffer;
+import com.ibm.j9ddr.corereaders.memory.MemoryFault;
+import com.ibm.j9ddr.corereaders.Platform;
+import com.ibm.j9ddr.corereaders.memory.AbstractMemory;
+import com.ibm.j9ddr.corereaders.memory.IMemorySource;
+import com.ibm.j9ddr.corereaders.memory.ProtectedMemoryRange;
+import java.nio.BufferUnderflowException;
+
+/**
+ * Serves as a ByteBuffer backed memory source.
+ *
+ * @see com.ibm.j9ddr.tools.ddrinteractive.CacheMemory
+ *
+ * @author knewbury01
+ *
+ */
+
+public class CacheMemorySource extends ProtectedMemoryRange implements IMemorySource
+{
+	private ByteBuffer bb;
+	protected final int asid;
+	protected final long baseAddress;
+	protected final long size;
+	
+	public CacheMemorySource(long address,
+							 ByteBuffer bb)
+	{
+		super(address,(long)bb.capacity());
+		this.asid = 0;
+		this.baseAddress = address;
+		this.size = (long)bb.capacity();
+		this.bb = bb;
+	}
+	
+	//
+	// Since the src is byte buffer backed
+	// it makes most sense to expose the following accessor
+	// methods from the src implementation
+	// as opposed to a getBytes (rewrapped) result 
+	//
+	public int getUnsignedShort(long address) {
+		int index = (address - this.baseAddress >>> 32) == 0 ? (int) (address - this.baseAddress) : -1;
```

Comment:
This range checking is not right. It should validate that all bytes to be accessed are in range of the given byte buffer. I suggest you want a helper method:
```
    private int checkIndex(long address, int byteCount) throws MemoryFault {
        if (byteCount < 0) {
            throw new IllegalArgumentException();
        }

        long maxAddress = baseAddress + Math.min(getSize(), Integer.MAX_VALUE) - byteCount;

        if ((address < baseAddress) || (address > maxAddress)) {
            throw new MemoryFault(address);
        }

        return (int) (address - baseAddress);
    }
```
that would be used like this:
```
    public short getShort(long address) throws MemoryFault {
        return source.getShort(checkIndex(address, Short.BYTES));
    }
```
Also note that these accessor methods should throw `MemoryFault` instead of `IndexOutOfBoundsException` or `BufferUnderflowException`.

Reply:
ah, had misunderstood, did not see that it was preferred to throw `MemoryFault`s instead of just letting them go to buffer exceptions